### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by goodboy**
Like it sounds a full suite of 10 tests to make sure this bad boy works!

Appreciate any feedback you guys might have :)

### CODE REVIEW

File: sandswitches/manage.py [hunk 1]
```diff
@@ -63,29 +82,99 @@ def __init__(self, rfile, etree, sftp, fscli, log):
...
+        allowed = ('domain', 'group', 'user', 'context')
+        if kwargs:
+            for argname, value in kwargs.items():
+                if argname not in allowed:
+                    raise ValueError(
+                        '{} is not a valid argument to list_users'.format(
+                            argname)
+                    )
+                args.append('{} {}'.format(argname, value))
+
```
Comment (angvp) [inline sandswitches/manage.py]: I would remove the ```if kwargs:``` condition, it's redundant, try to iterate directly .. if non extra arguments are passed then kwargs.items() is equals to []

File: sandswitches/orms.py [hunk 2]
```diff
@@ -148,17 +208,28 @@ class AttrMap(KeyValues):
...
+        return self.elem.xpath('/'.join((self.path, self.tag)))
+
+    @property
     def _attribs(self):
-        path = '/'.join((self.path, self.tag))
-        nodes = self.elem.xpath(path)
-        assert len(nodes) == 1, "Not an AttrMap?"
+        nodes = self._subelems
+        if len(nodes) == 0:  # no sub-elements exists yet
+            raise KeyError
```
Comment (vodik) [inline sandswitches/orms.py]: Is this an appropriate error to raise when there's no key?

File: sandswitches/orms.py [hunk 3]
```diff
@@ -169,7 +240,10 @@ def __getitem__(self, key):
     def __setitem__(self, key, value):
         if key in self.skipkeys:
             raise KeyError(key)
-        self._attribs[key] = value
+        try:
+            self._attribs[key] = value
+        except IndexError:
```
Comment (vodik) [inline sandswitches/orms.py:241]: Yeah, as per comment above, we can get either `KeyError` and `IndexError` out of `__setitem__` then, basically depending on the length?

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+
+@pytest.fixture(scope='session')
+def keycreds(request, fshost):
+    keyfile = request.config.option.keyfile
+    try:
+        sandswitches.comms.get_pkey(fshost, keyfile)
+        return keyfile, None
+    except Exception:
+        return keyfile, getpass.getpass("Password for {}: ".format(keyfile))
```
Comment (vodik) [inline tests/conftest.py]: Will this work nicely without suspending terminal capture?

File: sandswitches/comms.py [hunk 5]
```diff
@@ -18,7 +19,23 @@
...
+def get_pkey(keyfile, prompt=True, password=None):
+    """Get a private key references prompting for password by default.
+    """
+    try:
+        return paramiko.RSAKey.from_private_key_file(keyfile)
+    except paramiko.ssh_exception.PasswordRequiredException:
+        if not prompt and not password:
+            raise
+        if not password and prompt:
+            password = getpass("Password for {}: ".format(keyfile))
```
Comment (vodik) [inline sandswitches/comms.py]: Isn't this just:

```python
if not prompt:
    raise

password = getpass("Password for {}: ".format(keyfile))
return paramiko.RSAKey.from_private_key_file(keyfile, password=password)
```

File: tests/conftest.py (hunk 4 above)
Comment (vodik) [reply tests/conftest.py]: Actually, wait, doesn't `get_pkey` already have away to handle the prompt for you?

And could you not catch a more specific error?

File: sandswitches/comms.py (hunk 5 above)
Comment (goodboy) [reply sandswitches/comms.py]: Hmm yeah good point.

I was aiming for: `password` is provided and `prompt=False` -> attempt to use pw to get the key. 
Although, I guess I could just always pass `password=password` to `.from_private_key()` inside the `try:`.

File: sandswitches/manage.py (hunk 1 above)
Comment (goodboy) [reply sandswitches/manage.py]: Good call :)

File: tests/test_register_users.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    registrar['settings']['sip-port'] = str(
+        int(client['settings']['sip-port']) - 1)
+    registrar['domains']['test'] = {'parse': 'false', 'aliase': 'true'}
+    confmng.commit()
+    registrar.start()
+    client.start()
+
+    # register the client gateway to our 'doggy' user
+    client.register('doggy')
+    time.sleep(0.5)
```
Comment (goodboy) [inline tests/test_register_users.py]: Meant to drop this.

File: sandswitches/comms.py (hunk 5 above)
Comment (vodik) [reply sandswitches/comms.py]: Ahh, didn't notice that. That said, I suspect you'll be able to pass a password to a key that isn't password locked, and it'll still work.

Comment (goodboy) [reply sandswitches/comms.py]: Yeah the default value is `None`.

File: tests/conftest.py (hunk 4 above)
Comment (goodboy) [reply tests/conftest.py]: No to the first question, although I don't see why you'd try to run tests with an encrypted key if streams are being captured.

Yes `get_pkey` does do this but will prompt repeatedly due to multiple connections being created; one for each test. The idea here is that key creds are unlocked once at test session startup and then cached for further use.

Hhmm although I could probably just change the `confmng` fixture to be session scoped and maybe that willl do it. Let me try that out.

Comment (goodboy) [reply tests/conftest.py]: Hah yeah that worked.
Also makes the test run faster :)

File: sandswitches/orms.py (hunk 2 above)
Comment (goodboy) [reply sandswitches/orms.py]: Yeah this is for sure confusing and must have been from some kind of fix?
I can't recall atm why I did this.

Comment (goodboy) [reply sandswitches/orms.py]: Ahh yep from 13edd8f484b20396...
Yeah so looks like I can just drop `raise KeyError` stuff.

File: sandswitches/orms.py (hunk 3 above)
Comment (goodboy) [reply sandswitches/orms.py:241]: Yeah dropped the `KeyError` raise; was legacy.

**Comment by goodboy** [comment]
@vodik fixed everything I think.
Let me know :)

**Review by vodik — APPROVED** [review]
Looks good, nice work.

**Pull request merged**